### ISSUE TITLE
Protractor does not wait to resolve promises on routes

### ISSUE THREAD

**Issue by Zaylril** [angular/protractor#789]
Hey,

I'm having a few issues getting protractor to wait for my dependencies to load in the route.

The basic code I have is this (which works fine in Angular itself)

```
$routeProvider.when('/login, {
    templateUrl: 'template.html',
    controller: 'LoginCtrl',
    resolve: {
         dependencies: function($q, $rootScope) {
          var def = $q.defer();
          yepnope([
            {
                load: [
                    'dependencies.js',
                ],
                complete: function () {
                    $rootScope.$apply(function () {
                        def.resolve();
                    });
                }
            }
        ]);
        return def.promise;
        }
     }
 });
```

Protractor doesn't wait for the route to resolve and doesn't even display any content from the route. If the resolve dependency is removed the route successfully loads (but the test fails as the dependency isn't loaded).

I've tried it with $timeout testing and that works fine, and as this isn't a $http request, is there something I'm missing, or is this just not supported?

Thanks

**Comment by paulcwarren** [comment]
I am having exactly this issue on protrator 1.6.1.  Can we get some priority on this fix pls?  

In reality adding $timeout or browser.wait code is impractical.  Polluting production code to workaround a bug in the testing tool isnt really going to work.  But thanks for offering it.  Always useful to have something for the short-term.

This was first raised in May; i.e. been open a long time.  And this bug basically makes protractor useless for testing any real application since ng-route or ui-router are *always* going to be used and always going to be used with resolves (to avoid having to use ng-clock in partials which is another kludge).

So, +10 pls...

**Comment by facultymatt** [comment]
Im getting the same issue where resolves cause tests to fail, using ui-router v0.2.13.

**Comment by yxh10** [comment]
Guys please see the stackoverflow answer, that resolves it. You just need to ask Protractor to wait for the element to appear. Please see the code below 

'''javascript
browser.wait(function() {
       return $('#test321').isPresent(); // keeps waiting until this statement resolves to true
      }, 
      timeToWaitInMilliseconds, 
      'message to log to console if element is not present after that time'
    );

    expect($('#test321').isPresent()).toBe(true);  

'''

**Comment by Smolations** [comment]
@yxh10 That is not a long-term solution. Protractor should be aware of the views it's rendering, as well as when the DOM is ready. Using `browser.wait` in that fashion may solve the immediate problem, but if it happens across many tests, the spec files would be littered with all of those statements. Not only that, but how would one choose an appropriate `timeToWaitInMilliseconds`? Having multiple blocks such as the one you have above would significantly slow down testing for larger applications.

**Comment by yxh10** [comment]
@Smolations Hi, Chris, I agree with you. It is a work around. You could use this before protractor solve the problem from the core. 


Regarding to choose approriate ```timeToWaitInMilliseconds```, I am not sure if you are talking about the time you set in protractor config, or the time you set with browser.wait. If you are talking about the later, you actually don't need to worry about it. If you read the line that come with comment, and the api doc, you will realize it is returning a promise. The rest of the code will only run when the dom is presented (ready). It is exactly doing that. No time needed to be set. And if you want to wait till dom is actually displayed, you can use isDisplayed() instead of isPresent. 

So the only downside of this solution for now is too wordy for now. (You can always write your own util function to wrap it which is what I am doing in my current project.) You don't really need to worry about setting the length of time you wait. It is all handled by ```brower.wait``` waiting for the promise returned inside.

**Comment by moneytree-doug** [comment]
I am having this issue as well. As soon as I comment out my promises in the resolve of the ui-router, everything passes again.

**Comment by paulfreeman** [comment]
Also hitting problems with this today.  Work around makes testing very slow, but at least it works. Note, I believe that the timeToWaitInMilliseconds must be long enough to allow the actual request to return, browser.wait will then wait up to that length of time, but will return sooner if the request does.

**Comment by moneytree-doug** [comment]
I ended up using $httpBackend to stub out those requests in the resolve, but setting up $httpBackend is another story.

**Comment by yxh10** [comment]
@moneytree-doug , you could try the work around I posted above. But if your requirement is fine with mocked service, that is also cool. :+1:

**Comment by moneytree-doug** [comment]
@yxh10 I tried that, and it didn't work well. On top of that, I didn't want to throw an arbitrary number of seconds for waiting and prolong the test.

**Comment by yxh10** [comment]
@moneytree-doug Hm... what didn't work well? Can you specify? What browser are you using for the test? 

Protractor is developed to do End 2 End test. I think using mock sometimes is OK, but I think that defeat the original purpose. If you run against mock, just using Karma might be better?

**Comment by paulfreeman** [comment]
yxh10, I agree, with mocking you are doing another kind of testing, valid in its own way as an acceptance test of UX flow, but it isn't an end to end test.

**Comment by Smolations** [comment]
Not to get off topic here, but I think you guys (@yxh10 and @paulfreeman) have misunderstood what @moneytree-doug was trying to say. The way I read it, he had his original promises in place but was running into the alleged bug, which is the topic of this thread. He commented out his original code to verify that those promises were the issue, and was able to simulate their responses using `$httpBackend` in order to circumvent it. To me, he is most likely using protractor for proper e2e tests, but the problem is that _this apparent bug prevents normal e2e tests from passing_, regardless of whether his code is solid or not. I'm sure he tests his features manually during development to make sure the requests are properly formed, so mocked responses don't seem inappropriate to get around a potential bug. It seems like the available options are:

1. Wrap element references in tests with `$timeout`s, using an arbitrary number of milliseconds, thus delaying tests and incurring tech debt once this apparent bug is fixed (hopefully!)
2. Use mocked responses in the troubled route resolvers to ensure his tests will pass when valid data comes back from the server

E2E tests are meant to test organic user flows throughout an app. They are not necessarily used to test that a server returns valid responses, which is probably why he mocked the responses. Model/Unit tests can be used server-side to test the backend to ensure it behaves as it should.

Man, wish someone from the protractor team would jump in on this thread. Surely it's the most commented-on, open issue at this point!  :smile:

**Comment by paulfreeman** [comment]
@Smolations I see your point. I recognise that is the standard methodology. I still think that its sometimes useful to have a fully integrated e2e test and given the possible misnomer created by the term 'end to end' and the expectations it creates it would be a 'nice to have' idea to have this work with moderately realistic response times (say up to 10 seconds which I get from my iOS functional testing suite for example). Alternatively the documentation could point out the current behavior and the suggested methodology so that it helps a beginner.  For me I like to use a unit testing framework to test my backend API which I call an integration test, but I still see some value in aping the sequencing/timing of a realistic user scenario against a backend using the e2e test that can be run repetitively to simulate a realistic usage.

**Comment by moneytree-doug** [comment]
@Smolations Precisely!

**Comment by Smolations** [comment]
> Alternatively the documentation could point out the current behavior and the suggested methodology so that it helps a beginner.

Right?! Google releases best practices on almost everything else on the web, but it seems lacking for protractor. I try to keep a cool head though. The Angular/Protractor team is busy enough with 2.0 I'm sure, haha.

**Comment by moneytree-doug** [comment]
Out of curiosity here, what does the label "need investigation" mean? Does it need a set of steps to reproduce the issue? Or are we passed that and we are investigating where the bug is in protractor?

**Comment by krico** [comment]
I am working on an app that uses google javascript api (gapi) as the means to communicate with the backend (google cloud endpoints).  Unfortunately gapi works with 3rd party promises (neither $http or $timeout) and the proper way to handle this is to wrap those with `$q.when()` (please correct me if I'm wrong).
As a result the specs need to be written based on workarounds and are not reliable at all.
Here's an example of an imaginary blog service method to illustrate:
```javascript
function listBlogEntries() {
  return $q.when(gapi.client.myBlogApp.listBlogEntries());
}
```
If you call this method as a route resolve, or even at the click of a button, protractor doesn't wait for it to "finish".  This means that the tests need to either "sleep" and hope for the best, or you end up having to add elements to your app.  So from my side, I give this issue a big :+1: and also volunteer to help ;-)

**Comment by sjelin** [comment]
@hankduan something to keep in mind for the testability API in Angular2?

**Comment by sjelin** [comment]
Minimal example:

```html
<html ng-app="countryApp">
  <head>
    <meta charset="utf-8">
    <title>Angular.js Example</title>
    <script src="angular.js"></script>
    <script src="angular-route.js"></script>
    <script>
      var countryApp = angular.module('countryApp', ['ngRoute']);

      countryApp.config(function($routeProvider) {
        $routeProvider.
          when('/', {
            template: '<ul><li ng-repeat="country in countries">{{country.name}}</li><ul>',
            controller: 'CountryListCtrl',
            resolve: {
              countries: function($q) {
                var deferred = $q.defer();
                setTimeout(function() {
                  deferred.resolve([{
                    "name": "China",
                    "population": 1359821000,
                    "flagURL": "//upload.wikimedia.org/wikipedia/commons/f/fa/Flag_of_the_People%27s_Republic_of_China.svg",
                    "capital": "Beijing",
                    "gdp": 12261
                  }, {
                    "name": "India",
                    "population": 1205625000,
                    "flagURL": "//upload.wikimedia.org/wikipedia/en/4/41/Flag_of_India.svg",
                    "capital": "New Delhi",
                    "gdp": 4716
                  }, {
                    "name": "United States of America",
                    "population": 312247000,
                    "flagURL": "//upload.wikimedia.org/wikipedia/en/a/a4/Flag_of_the_United_States.svg",
                    "capital": "Washington, D.C.",
                    "gdp": 16244
                  }]);
                }, 3000);
                return deferred.promise;
              }
            }
          }).
          when('/:countryName', {
            template: '<h1>TODO create country detail view</h1>',
            controller: 'CountryDetailCtrl'
          }).
          otherwise({
            redirectTo: '/'
          });
      });

      countryApp.controller('CountryListCtrl', function ($scope, $http){
        $http.get('countries.json').success(function(data) {
          $scope.countries = data;
        });
      });

      countryApp.controller('CountryDetailCtrl', function ($scope, $routeParams){
        console.log($routeParams);
      });
    </script>
  </head>
  <body>
    <div ng-view></div>
  </body>
</html>
```

```js
describe('', function() {
  it('', function() {
    browser.get('index.html');
    expect(element.all(by.tagName('li')).count()).toBe(3);
  });
});
```

Will have PR to angular team up in a few minutes

**Comment by fgather** [comment]
I see the same problem with Angular2 beta16 when using routes which load some additional data from the back end. Did someone already investigate this or am I the only one seeing this problem?

**Comment by juliemr** [comment]
@sjelin is there a link to a PR for angular?

**Comment by gkalpak** [comment]
@juliemr , I think @sjelin meant angular/angular.js#14159.

**Comment by gkalpak** [comment]
I think this can be closed now that angular/angular.js#14159 has been merged, but leaving it up to you :smiley:
(The fix will be included in v1.6.2 btw.)

cc @sjelin, @juliemr
